samples/zigbee/light_bulb/src/zcl_scenes.c

```diff
@@ -0,0 +1,929 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "zcl_scenes.h"
+#include <zb_nrf_platform.h>
+#include <logging/log.h>
+#include <settings/settings.h>
+
+#define LOG_MODULE_NAME zcl_scenes
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+static zb_uint8_t scene_table_get_entry(zb_uint16_t group_id, zb_uint8_t scene_id);
+static void scene_table_remove_entries_by_group(zb_uint16_t group_id);
+static void scene_table_init(void);
+
+struct zb_zcl_scenes_fieldset_data_on_off {
+	zb_bool_t  has_on_off;
+	zb_uint8_t on_off;
+};
+
+struct zb_zcl_scenes_fieldset_data_level_control {
+	zb_bool_t  has_current_level;
+	zb_uint8_t current_level;
+};
+
+struct zb_zcl_scenes_fieldset_data_window_covering {
+	zb_bool_t  has_current_position_lift_percentage;
+	zb_uint8_t current_position_lift_percentage;
+	zb_bool_t  has_current_position_tilt_percentage;
+	zb_uint8_t current_position_tilt_percentage;
+};
+
+struct scene_table_on_off_entry {
+	zb_zcl_scene_table_record_fixed_t                  common;
+	struct zb_zcl_scenes_fieldset_data_on_off          on_off;
+	struct zb_zcl_scenes_fieldset_data_level_control   level_control;
+	struct zb_zcl_scenes_fieldset_data_window_covering window_covering;
+};
+
+struct scene_table_on_off_entry scenes_table[ZCL_SCENES_TABLE_SIZE];
+
+struct response_info {
+	zb_zcl_parsed_hdr_t cmd_info;
+	zb_zcl_scenes_view_scene_req_t view_scene_req;
+	zb_zcl_scenes_get_scene_membership_req_t get_scene_membership_req;
+};
+
+struct response_info resp_info;
+
+static int scenes_table_set(const char *name, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	const char *next;
+	int rc;
+
+	if (settings_name_steq(name, "scenes_table", &next) && !next) {
+		if (len != sizeof(scenes_table)) {
+			return -EINVAL;
+		}
+
+		rc = read_cb(cb_arg, scenes_table, sizeof(scenes_table));
+		if (rc >= 0) {
+			return 0;
+		}
+
+		return rc;
+	}
+
+	return -ENOENT;
+}
+
+static void scenes_table_save(void)
+{
+	settings_save_one("scenes/scenes_table", scenes_table, sizeof(scenes_table));
+}
+
+struct settings_handler scenes_conf = {
+	.name = "scenes",
+	.h_set = scenes_table_set
+};
+
+static zb_bool_t has_cluster(zb_uint16_t cluster_id)
+{
+	return (get_endpoint_by_cluster(cluster_id, ZB_ZCL_CLUSTER_SERVER_ROLE)
+		== ZCL_SCENES_ENDPOINT)
+		? ZB_TRUE : ZB_FALSE;
+}
+
+static zb_bool_t add_fieldset(zb_zcl_scenes_fieldset_common_t *fieldset,
+			      struct scene_table_on_off_entry *p_entry)
+{
+	zb_uint8_t *fs_data_ptr;
+
+	if (fieldset->cluster_id == ZB_ZCL_CLUSTER_ID_ON_OFF &&
+	    fieldset->fieldset_length >= 1 &&
+	    has_cluster(ZB_ZCL_CLUSTER_ID_ON_OFF)) {
+		fs_data_ptr = (zb_uint8_t *)fieldset + sizeof(zb_zcl_scenes_fieldset_common_t);
+		p_entry->on_off.has_on_off = ZB_TRUE;
+		p_entry->on_off.on_off = *fs_data_ptr;
+
+		LOG_INF("Add fieldset: cluster_id=0x%x, length=%d, on/off=%d",
+			fieldset->cluster_id, fieldset->fieldset_length, p_entry->on_off.on_off);
+		return ZB_TRUE;
```

Comment:
Could we add some empty lines to improve readability of this code block?

Reply:
I added a few empty lines in this `if` block and in the following `if` blocks in this function. Besides, I removed all else's because of checkpatch suggestion. It should be more readable now.